README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# dacol

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/ldanai/dacol.svg?branch=master)](https://travis-ci.org/ldanai/dacol)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/ldanai/dacol?branch=master&svg=true)](https://ci.appveyor.com/project/ldanai/dacol)
[![Codecov test
coverage](https://codecov.io/gh/ldanai/dacol/branch/master/graph/badge.svg)](https://codecov.io/gh/ldanai/dacol?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/dacol)](https://cran.r-project.org/package=dacol)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- [![DOI](https://zenodo.org/badge/123997166.svg)](https://zenodo.org/badge/latestdoi/123997166) -->
<!-- badges: end -->

dacol provides utilities to add or modify columns in dataframe.

The utilities include:

  - Statistical measures: `mode`, `confident_interval`, `ceiling`
  - Normalize a vector column: `cosine`, `logistic`, `zscore`
  - Compute distance between 2 vector columns: `euclidean`, `pearson`,
    `cosine`, `canberra`
  - Manage outliers: `trim_outlier`, `normalize_ptile`
  - Calculate percentile: `decile_band`, `decile_ptile`, `dc_rank_ptile`

More info: <https://ldanai.github.io/dacol/>

## Installation

You can install dacol from github with:

``` r
# install.packages("remotes")
remotes::install_github("ldanai/dacol")
```

## Example

This shows how to use dacol:

``` r
library(dacol)
library(dplyr)

max = 30
dta1 = tibble(x1 = seq(-1.2*max, 1.2*max, length.out = 200),
              x2 = seq(0, max, length.out = 200),
              x3 = sample(200))

dta1
#> # A tibble: 200 x 3
#>       x1    x2    x3
#>    <dbl> <dbl> <int>
#>  1 -36   0       102
#>  2 -35.6 0.151    18
#>  3 -35.3 0.302    72
#>  4 -34.9 0.452   148
#>  5 -34.6 0.603   194
#>  6 -34.2 0.754    40
#>  7 -33.8 0.905   131
#>  8 -33.5 1.06     66
#>  9 -33.1 1.21     21
#> 10 -32.7 1.36    170
#> # ... with 190 more rows

dta1 = 
  dta1 %>% 
  mutate(
    # Transformation
    y_cosine   = dc_cosine(x1, max),
    y_logistic = dc_logistic(x2, max),
    y_zcore    = dc_zscore(x2),
    
    # Distant between 2 vector columns
    y_dist_canb = dc_dist_canberra(x2, x3),
    y_dist_cos  = dc_dist_cosine(x2, y_zcore),
    y_dist_euc  = dc_dist_euclidean(x2, y_zcore),
    y_dist_pear = dc_dist_pearson(x2, y_zcore),
    
    # Manage outliers
    y_trim = dc_trim_outlier(x3, 0.01),
    y_norm = dc_normalize_ptile(x3, 0.01),
    
    # Stats measures
    y_mode = dc_mode(x3),
    y_ceil = dc_ceiling(x1, -1),
    
    # Band segmentation
    y_dec_band1  = dc_decile_band(x3),
    y_dec_band2  = dc_decile_band(x3, c(seq(0, 0.9, 0.1))),
    y_dec_ptile1 = dc_decile_ptile(x3),
    y_dec_ptile2 = dc_decile_ptile(x3, c(seq(0, 0.9, 0.1))),
    
    # Rank percentile
    y_ranked1 = dc_rank_ptile(x3), 
    y_ranked2 = dc_rank_ptile(x3, c(seq(1, 100, 1))) 
  )

dta1
#> # A tibble: 200 x 20
#>       x1    x2    x3 y_cosine y_logistic y_zcore y_dist_canb y_dist_cos
#>    <dbl> <dbl> <int>    <dbl>      <dbl>   <dbl>       <dbl>      <dbl>
#>  1 -36   0       102        0     0        -1.72     0            0.498
#>  2 -35.6 0.151    18        0     0.0251   -1.70     0.00831      0.498
#>  3 -35.3 0.302    72        0     0.0502   -1.68     0.00417      0.498
#>  4 -34.9 0.452   148        0     0.0752   -1.67     0.00305      0.498
#>  5 -34.6 0.603   194        0     0.100    -1.65     0.00310      0.498
#>  6 -34.2 0.754    40        0     0.125    -1.63     0.0185       0.498
#>  7 -33.8 0.905   131        0     0.150    -1.62     0.00686      0.498
#>  8 -33.5 1.06     66        0     0.174    -1.60     0.0157       0.498
#>  9 -33.1 1.21     21        0     0.198    -1.58     0.0543       0.498
#> 10 -32.7 1.36    170        0     0.222    -1.56     0.00792      0.498
#> # ... with 190 more rows, and 12 more variables: y_dist_euc <dbl>,
#> #   y_dist_pear <dbl>, y_trim <dbl>, y_norm <dbl>, y_mode <int>,
#> #   y_ceil <dbl>, y_dec_band1 <dbl>, y_dec_band2 <dbl>,
#> #   y_dec_ptile1 <dbl>, y_dec_ptile2 <dbl>, y_ranked1 <dbl>,
#> #   y_ranked2 <dbl>
```

-----

Please note that the ‘dacol’ project is released with a [Contributor
Code of Conduct](.github/CODE_OF_CONDUCT.md). By contributing to this
project, you agree to abide by its terms.


NEWS.md
# dacol (development version)

* Add new function `cd_ceiling()`
* Add a `NEWS.md` file to track changes to the package.

# dacol 0.2

* Use Travis CI with `use_tidy_ci()` and `use_travis()`
* Add rank_ptile: a function to rank in percentile



tests/testthat/test-transformation.R
context("Data transformation")

test_that("Basic dc_cosine works", {
  expect_equal(dc_cosine(0), 1)
  expect_equal(dc_cosine(0, 10), 1)

  expect_lt(dc_cosine(1), 1)
  expect_lt(dc_cosine(-1), 1)

  expect_lt(dc_cosine(9, 10), 1)
  expect_lt(dc_cosine(-9, 10), 1)

  expect_equal(dc_cosine(1e10, 1), 0)
  expect_equal(dc_cosine(-1e10, 1), 0)
})


test_that("Basic dc_logistic works", {
  expect_equal(dc_logistic(0), 0)
  expect_equal(dc_logistic(0, 10), 0)

  expect_lt(dc_logistic(1), 1)
  expect_lt(dc_logistic(-1), 1)

  expect_lt(dc_logistic(9, 10), 1)
  expect_lt(dc_logistic(-9, 10), 1)

  expect_equal(dc_logistic(1e5, 1), 1)
  expect_equal(dc_logistic(-1e5, 1), -1)
})


test_that("Basic dc_zscore works", {
  x = seq(1, 10)
  expect_equal(length(dc_zscore(x)), length(x))
  expect_equal(mean(dc_zscore(x)), 0)
  expect_equal(sd(dc_zscore(x)), 1)
})

test_that("Basic dc_trim_outlier works", {
  x = c(seq(1, 10), 1000)
  expect_error(dc_trim_outlier(data.frame(a=1:2, b=1:2)), "x must be vector")
  expect_equal(length(dc_trim_outlier(x)), length(x))
})


test_that("Basic dc_dist_canberra works", {
  expect_error(dc_dist_canberra(-1, 1), "x must be non negative value")
  expect_error(dc_dist_canberra(1, -1), "y must be non negative value")
  expect_error(dc_dist_canberra(-1, -1), "x must be non negative value")

  expect_equal(dc_dist_canberra(0, 0), NaN)

  expect_equal(dc_dist_canberra(0, 1e10), 0)
  expect_equal(dc_dist_canberra(1e10, 0), 1)
  expect_equal(dc_dist_canberra(1, 1e10), 0)
  expect_lt(dc_dist_canberra(1e10, 1), 1)

  expect_lt(dc_dist_canberra(1, 2), 1)
  expect_lt(dc_dist_canberra(10, 4), 1)
})

test_that("Basic dc_dist_cosine works", {
  expect_equal(dc_dist_cosine(c(1, 0), c(1, 0)), 0)
  expect_equal(dc_dist_cosine(c(1, 0), c(0, 1)), 1)

  expect_error(dc_dist_cosine(c(1, 1, 1), c(2, 2)), "x and y must have the same length vectors")
  expect_error(dc_dist_cosine(data.frame(a=1:2, b=1:2), c(1, 0)), "x must be vector")
  expect_error(dc_dist_cosine(c(1, 0), data.frame(a=1:2, b=1:2)), "y must be vector")
})

test_that("Basic dc_dist_euclidean works", {
  expect_equal(dc_dist_euclidean(c(1, 0), c(1, 0)), 0)
  expect_equal(dc_dist_euclidean(c(1, 0), c(0, 1)), sqrt(2))
  expect_equal(dc_dist_euclidean(c(1, 0), c(0, -1)), sqrt(2))
  expect_equal(dc_dist_euclidean(c(-1, 0), c(0, -1)), sqrt(2))

  expect_error(dc_dist_euclidean(c(1, 1, 1), c(2, 2)), "x and y must have the same length vectors")
  expect_error(dc_dist_euclidean(data.frame(a=1:2, b=1:2), c(1, 0)), "x must be vector")
  expect_error(dc_dist_euclidean(c(1, 0), data.frame(a=1:2, b=1:2)), "y must be vector")
})

test_that("Basic dc_dist_pearson works", {
  expect_equal(dc_dist_pearson(c(1, 0), c(0, 1)), 1)
  expect_equal(dc_dist_pearson(c(1, 0), c(-1, 0)), 1)
  expect_lt(dc_dist_pearson(c(1, 0), c(1, 0)), 1e-10)
  expect_lt(dc_dist_pearson(c(0, 1), c(0, 1)), 1e-10)

  expect_error(dc_dist_pearson(c(1, 1, 1), c(2, 2)), "x and y must have the same length vectors")
  expect_error(dc_dist_pearson(data.frame(a=1:2, b=1:2), c(1, 0)), "x must be vector")
  expect_error(dc_dist_pearson(c(1, 0), data.frame(a=1:2, b=1:2)), "y must be vector")
})


test_that("Basic dc_normalize_ptile works", {
  expect_error(dc_normalize_ptile(data.frame(a=1:2, b=1:2)), "x must be vector")

  x = c(seq(1, 10), 1000)
  expect_equal(length(dc_normalize_ptile(x)), length(x))
  expect_equal(min(dc_normalize_ptile(x)), -1)
  expect_equal(max(dc_normalize_ptile(x)), 1)
})


test_that("Basic get_confidence_interval works", {
  expect_error(get_confidence_interval(data.frame(a=1:2, b=1:2)), "x must be vector")
  expect_error(get_confidence_interval(seq(1, 10), level=2), "level must be between 0 and 1")

  x = c(seq(1, 10), 1000)
  expect_equal(get_confidence_interval(x)$mean, mean(x))
})


test_that("Basic dc_decile_band works", {
  expect_error(dc_decile_band(data.frame(a=1:2, b=1:2)), "x must be vector")

  x = c(seq(1, 10), 1000)
  expect_equal(length(dc_decile_band(x)), length(x))
})



test_that("Basic dc_decile_ptile works", {
  expect_error(dc_decile_ptile(data.frame(a=1:2, b=1:2)), "x must be vector")

  x = c(seq(1, 10), 1000)
  expect_equal(length(dc_decile_ptile(x)), length(x))
})

test_that("Basic dc_rank_ptile works", {
  expect_error(dc_rank_ptile(data.frame(a=1:2, b=1:2)), "x must be vector")

  x = c(seq(1, 10), 1000)
  expect_equal(length(dc_rank_ptile(x)), length(x))
})

test_that("Basic dc_mode works", {
  expect_error(dc_mode(data.frame(a=1:2, b=1:2)), "x must be vector")

  x = c(1, 2, 3, 3, 4, 5)
  expect_equal(length(dc_mode(x)), 1)
  expect_equal(dc_mode(x), 3)
  expect_equal(dc_mode(c(1, 2, 3, 4, 5)), 1)

  expect_equal(dc_mode(c(1, 2, 3, 3, NA, 4, 5), na.rm = TRUE), 3)
})

# test_that("Basic mode_stats works", {
#   expect_error(mode_stats(data.frame(a=1:2, b=1:2)), "x must be vector")
#
#   x = c(1, 2, 3, 3, 4, 5)
#   expect_equal(length(mode_stats(x)), 1)
#   expect_equal(mode_stats(x), 3)
#   expect_equal(mode_stats(c(1, 2, 3, 4, 5)), 1)
#
#   expect_equal(mode_stats(c(1, 2, 3, 3, NA, 4, 5), na.rm = TRUE), 3)
# })

test_that("Basic dc_ceiling works", {
  expect_error(dc_ceiling(data.frame(a=1:2, b=1:2)), "x must be vector")

  x = c(5, 21, 39, 31, 40, 15)
  expect_equal(length(dc_ceiling(x)), 6)
  expect_equal(dc_ceiling(x, -1), c(10, 30, 40, 40, 40, 20))
  expect_equal(dc_ceiling(x*10, -2), c(100, 300, 400, 400, 400, 200))

  expect_equal(dc_ceiling(c(1, 2, 3, 3, NA, 4, 5), na.rm = TRUE), c(2, 2, 4, 4, 4, 6))
})




R/dacol.R
#' dacol: Danai's utilities for adding columns
#'
#' dacol provides utilities to add/modify columns into dataframe.
#'
#' The utilities include:
#'
#' \itemize{
#' \item Statistical measures: `mode`, `confident_interval`, `dc_ceiling`
#' \item Normalize a vector column: `cosine`, `logistic`, `zscore`
#' \item Compute distance between 2 vector columns: `euclidean`, `pearson`, `cosine`, `canberra`
#' \item Manage outliers: `trim_outlier`, `normalize_percentile`
#' \item Calculate percentile: `decile_band`, `decile_ptile`, `rank_ptile`
#' }
#'
#' To learn more about dplyr, start with the vignettes:
#' `browseVignettes(package = "dacol")`
#'
#' @section Package options:
#' \describe{
#' \item{`dacol.show_progress`}{Should lengthy operations such as `do()`
#'   show a progress bar? Default: `TRUE`}
#' }
#'
#' @import rlang
#' @import dplyr
#' @import tidyr
#' @import ggplot2
#' @importFrom glue glue
#' @importFrom stats qt
#' @importFrom stats sd
#' @importFrom stats quantile
#' @importFrom stats reorder
#'
"_PACKAGE"
NULL
utils::globalVariables(c("."))


R/transformation.R
#' Functions to normalize, transform, measure distance between numeric vectors
#'
#' \code{dc_cosine} is the cosine transformation.
#' \code{dc_logistic} is the logistic transformation.
#' \code{dc_zscore} is the zscore transformation.
#' \code{dc_dist_canberra} computes the Canberra distance between 2 numeric vectors.
#' \code{dc_dist_cosine} computes the cosine angle distance between 2 numeric vectors.
#' \code{dc_dist_euclidean} compute the Euclidience distance between 2 numeric vectors.
#' \code{dc_dist_pearson} compute the Pearson correlation distance between 2 numeric vectors.
#'
#' \code{dc_ceiling} similar to rbase::ceiling() with support decimal round up
#' \code{dc_mode} compute the stats mode
#'
#' \code{dc_rank_ptile} add columns with ranked percentiles
#' \code{dc_decile_band} add columns with decile bands
#' \code{dc_decile_ptile} add columns with decile percentiles
#'
#' @param x A numeric vector
#' @param y A numeric vector
#' @param max A numeric value
#' @param fraction The percentile value (0 to 0.5) to trim out
#' @param level The CI level (0.5 to 1.0) of observations to be measured.
#' @param band_ptile The percentail band (0.0 to 1.0)
#' @param level_rank The rank level (0.0 to 1.0) for calculating percentile
#' @param na.rm	A logical value indicating whether NA values should be stripped before the computation proceeds.
#' @param digits similar to rbase::round() which is integer indicating the number of decimal places (round) or significant digits (signif) to be used. Negative values are allowed
#'
#' @return returns a numeric vector after normaliztion or distance between 2 vectors.
#'
#' @name data-normalization
#'
#' @examples
#'
#' library(dacol)
#' library(dplyr)
#'
#' max = 30
#' dta1 = tibble(x1 = seq(-1.2*max, 1.2*max, length.out = 200),
#'               x2 = seq(1, max, length.out = 200),
#'               x3 = sample(200))
#'
#' dta1 = mutate(dta1,
#'
#'               # Transformation
#'               y_cosine   = dc_cosine(x1, max),
#'               y_logistic = dc_logistic(x2, max),
#'               y_zcore    = dc_zscore(x2),
#'
#'               # Distant between 2 vector columns
#'               y_dist_canb = dc_dist_canberra(x2, x3),
#'               y_dist_cos  = dc_dist_cosine(x2, y_zcore),
#'               y_dist_euc  = dc_dist_euclidean(x2, y_zcore),
#'               y_dist_pear = dc_dist_pearson(x2, y_zcore),
#'
#'               # Manage outliers
#'               y_trim = dc_trim_outlier(x3, 0.01),
#'               y_norm = dc_normalize_ptile(x3, 0.01),
#'
#'               # Stats measures
#'               y_mode = dc_mode(x3),
#'               y_ceil = dc_ceiling(x3, -1),
#'
#'               # Band segmentation
#'               y_dec_band1 = dc_decile_band(x3),
#'               y_dec_band2 = dc_decile_band(x3, c(seq(0, 0.9, 0.1))),
#'               y_dec_ptile1 = dc_decile_ptile(x3),
#'               y_dec_ptile2 = dc_decile_ptile(x3, c(seq(0, 0.9, 0.1)))
#'               )
NULL

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_cosine = function(x, max = 100)
{
  value = 0.5*(1 + cos((pi/max)*x))

  value = ifelse(abs(x) > max, 0, value)
  value
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_logistic = function(x, max = 100)
{
  a     = 1/(0.1*max)             # 0.1*max --> at y = 0.5
  value = 2/(1 + exp(- a*x)) - 1  # (x > 0) --> y in [0, 1]
  value
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_zscore = function(x)
{
  return((x-mean(x))/sd(x))
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_dist_canberra = function(x, y)
{
  if(sum(x < 0) > 0) abort("x must be non negative value")
  if(sum(y < 0) > 0) abort("y must be non negative value")

  return(0.5*(1 + (x-y)/(x+y)))
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_dist_cosine = function(x, y)
{
  if(!is.vector(x)) abort("x must be vector")
  if(!is.vector(y)) abort("y must be vector")
  if(length(x) != length(y)) abort("x and y must have the same length vectors")

  return(1 - sum(x*y)/(sqrt(sum(x*x))*sqrt(sum(y*y))))
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_dist_euclidean = function(x, y)
{
  if(!is.vector(x)) abort("x must be vector")
  if(!is.vector(y)) abort("y must be vector")
  if(length(x) != length(y)) abort("x and y must have the same length vectors")

  return(sqrt(sum((x-y)^2)))
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_dist_pearson = function(x, y)
{
  if(!is.vector(x)) abort("x must be vector")
  if(!is.vector(y)) abort("y must be vector")
  if(length(x) != length(y)) abort("x and y must have the same length vectors")

  return(0.5*(1 - (sum(x*y)-length(x)*mean(x)*mean(y))/((length(x)-1)*sd(x)*sd(y))))
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_trim_outlier = function(x, fraction=0.01)
{
  if(!is.vector(x)) abort("x must be vector")

  threshold.low  = quantile(x, fraction, na.rm = TRUE)
  threshold.high = quantile(x, 1-fraction, na.rm = TRUE)

  x[x<=threshold.low]  = threshold.low
  x[x>=threshold.high] = threshold.high

  return(x)
}


###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_normalize_ptile = function(x, fraction = 0.01)
{
  if(!is.vector(x)) abort("x must be vector")

  x = dc_trim_outlier(x, fraction)
  x = (x-min(x))/(max(x)-min(x))   #Scale data to [0, 1]
  x = 2*x-1                        #Scale all values in range [-1, 1]
  return(x)
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
get_confidence_interval = function(x, level=0.95)
{
  if(!is.vector(x)) abort("x must be vector")

  if (level <= 0 || level>=1)
    abort("level must be between 0 and 1")

  m  = mean(x)
  n  = length(x)
  SE = sd(x)/sqrt(n)
  upper = 1 - (1-level)/2
  ci = m + c(-1,1)*qt(level, n-1)*SE

  list(mean = m,
       se   = SE,
       ci   = ci)
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_decile_band = function(x, band_ptile = c(seq(0, 0.95, 0.05)))
{
  if(!is.vector(x)) abort("x must be vector")

  band_decile = quantile(x, probs = band_ptile)
  idx         = findInterval(x, band_decile)

  # return
  band_decile[idx]
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_decile_ptile = function(x, band_ptile = c(seq(0, 0.95, 0.05)))
{
  if(!is.vector(x)) abort("x must be vector")

  band_decile = quantile(x, probs = band_ptile)
  idx         = findInterval(x, band_decile)

  # return
  band_ptile[idx]
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_rank_ptile = function(x, level_rank = c(1, 2, 3, 4, seq(5, 100, 5)))
{
  if(!is.vector(x)) abort("x must be vector")

  level_rank   = sort(level_rank, decreasing = TRUE)
  pct_interval = quantile(x, probs = 1 - level_rank/100)
  level_rank[findInterval(x, pct_interval)]
}

###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_mode = function(x, na.rm = FALSE) {

  if(!is.vector(x)) abort("x must be vector")

  if(na.rm){
    x = x[!is.na(x)]
  }

  ux = unique(x)
  return(ux[which.max(tabulate(match(x, ux)))])
}


###-----------------------------------------------------------------------------
#' @export
#' @rdname data-normalization
###-----------------------------------------------------------------------------
dc_ceiling = function(x, digits = 0, na.rm = FALSE) {

  if(!is.vector(x)) abort("x must be vector")

  if(na.rm){
    x = x[!is.na(x)]
  }

  a = (10^(-digits))/2 - 1e-100

  return(round(x + a, digits))
}
